pkg/synth/acl.go

```diff
@@ -27,156 +28,70 @@ func NewACLSynthesizer(s *ir.Spec, single bool) Synthesizer {
 	return &ACLSynthesizer{spec: s, singleACL: single, result: ir.NewACLCollection()}
 }
 
-func (a *ACLSynthesizer) Synth() (ir.Collection, error) {
+func (a *ACLSynthesizer) Synth() ir.Collection {
 	return a.makeACL()
 }
 
 // makeACL translates Spec to a collection of nACLs
 // 1. generate nACL rules for relevant subnets for each connection
 // 2. generate nACL rules for blocked subnets (subnets that do not appear in Spec)
-func (a *ACLSynthesizer) makeACL() (*ir.ACLCollection, error) {
-	for c := range a.spec.Connections {
-		err := a.generateACLRulesFromConnection(&a.spec.Connections[c])
-		if err != nil {
-			return nil, err
-		}
+func (a *ACLSynthesizer) makeACL() *ir.ACLCollection {
+	for i := range a.spec.Connections {
+		conn := a.spec.Connections[i]
+		a.generateACLRulesFromConnection(conn, conn.Src, conn.Dst, a.allowConnectionSrc)
+		a.generateACLRulesFromConnection(conn, conn.Dst, conn.Src, a.allowConnectionDst)
 	}
 	a.generateACLRulesForBlockedSubnets()
-	return a.result, nil
+	return a.result
 }
 
-//  1. check that both resources are supported in nACL generation.
-//  2. check that at least one resource is internal.
-//  3. convert src and dst resources to namedAddrs slices to make it more convenient to go through the addrs
-//     and add the rule to the relevant acl. Note: in case where the resource in a nif, src/dst will be
-//     updated to be its subnet.
-//  4. generate rules and add them to relevant ACL to allow traffic for all pairs of IPAddrs of both resources.
-func (a *ACLSynthesizer) generateACLRulesFromConnection(conn *ir.Connection) error {
-	if !resourceRelevantToACL(conn.Src.Type) {
-		return fmt.Errorf(ACLTypeNotSupported, string(conn.Src.Type))
-	}
-	if !resourceRelevantToACL(conn.Dst.Type) {
-		return fmt.Errorf(ACLTypeNotSupported, string(conn.Dst.Type))
-	}
-	for _, src := range conn.Src.IPAddrs {
-		srcSubnets, srcCidr := adjustResource(&a.spec.Defs, src, conn.Src)
-		for _, dst := range conn.Dst.IPAddrs {
-			dstSubnets, dstCidr := adjustResource(&a.spec.Defs, dst, conn.Dst)
-			if src == dst && conn.Src.Type != ir.ResourceTypeCidr && conn.Dst.Type != ir.ResourceTypeCidr {
+func (a *ACLSynthesizer) generateACLRulesFromConnection(conn *ir.Connection, thisResource, otherResource *ir.LocalRemotePair,
+	allowConnection func(*ir.Connection, *ir.TrackedProtocol, *ir.NamedAddrs, *netset.IPBlock)) {
+	for _, thisSubnet := range thisResource.LocalCidrs {
+		for _, otherCidr := range otherResource.RemoteCidrs {
+			if thisSubnet.IPAddrs.Equal(otherCidr.IPAddrs) && thisResource.LocalType != ir.ResourceTypeCidr &&
```

Comment:
Seems like we can remove the type checks in this if.
It also seems like this check is repeated in the `allowConnection*` functions.